// thumbnails-webvtt/lib/writer/spritesheet.js
'use strict'

var fs = require('fs-extra')
  , lodash = require('lodash')
  , path = require('path')
  , sprite = require('node-sprite')
  , util = require('util')
  , utils = require('../utils')
  , Writer = require('../writer')




/**
 * Small helper to be sure we're not getting EEXIST
 */
function unlinkAndMove(src, dst, callback) {
  fs.exists(dst, onexists)

  function onexists(exists) {
    if (!exists) {
      move(null)
    }
    else {
      fs.unlink(dst, move)
    }
  }

  function move(err) {
    if (err) {
      callback(err)
    }
    else {
      fs.move(src, dst, callback)
    }
  }
}




/**
 * Creates spritesheet then writes files
 *
 * @constructor
 * @extends {Writer}
 */
function SpriteSheetWriter(metadata, options, filenames) {
  Writer.call(this, metadata, options, filenames)
  var self = this
    , spriteOptions
    , sheetName = options.spriteSheetName || 'thumbnails'
    , spritesDirectory = path.join(options.thumbnailsDirectory, sheetName)
    , paths = []
    , jsonMetaDataArray = []

  spriteOptions = lodash.extend({
    path: options.thumbnailsDirectory
  }, options.spriteSheetOptions || {})

  fs.exists(spritesDirectory, onexists)

  function onexists(exists) {
    if (exists) {
      onmkdir(null)
    }
    else {
      fs.mkdir(spritesDirectory, onmkdir)
    }
  }


  function onmkdir(err) {
    if (err) {
      return self.emit('internalError', err)
    }

    var tomove = filenames.length
      , error = null
      , i = 0
      , thumbnail

    while (thumbnail = filenames[i++]) {
      move(thumbnail)
    }

    function move(thumbnail) {
      var src = path.join(options.thumbnailsDirectory, thumbnail)
        , dst = path.join(spritesDirectory, thumbnail)

      unlinkAndMove(src, dst, ondone)
    }

    function ondone(err) {
      if (error) {
        return
      }
      else if (err) {
        error = err
        self.emit('internalError', err)
      }
      else if (!--tomove) {
        try{createSpriteSheet()}catch(err){}
      }
    }
  }

  function createSpriteSheet() {
    sprite = require('node-sprite');
    sprite.sprite(sheetName, spriteOptions, function(err, globalSprite) {
      if (err) {
        return self.emit('internalError', err)
      }

      var sprites = globalSprite.images
        , sheetFilename = globalSprite.filename()
        , sheetExt = path.extname(sheetFilename)
        , filename
        , i = 0
        , spritePath
        , jsonMetaData

      while (filename = filenames[i++]) {
        sprite = lodash.find(sprites, {filename: filename})


        if(sprite){
          //TODO: Hier wird der path geschrieben

        jsonMetaData = {
          posX : sprite.positionX,
          posY : sprite.positionY,
          width : sprite.width,
          height : sprite.height,
          fps : self.metadata.fps
        }
        spritePath = util.format(
          '/%s/%s#xywh=%d,%d,%d,%d',
          options.assetsDirectory,
          sheetName + sheetExt,
          sprite.positionX,
          sprite.positionY,
          sprite.width,
          sprite.height
        );
        } else {
          var e = new Error ('spritesheet.js::createSpritesheet::sprite=undefined');
          return self.emit('internalError', e);
        }

        paths.push(spritePath);
        jsonMetaDataArray.push(jsonMetaData);
      }

      renameSheetName(sheetFilename, sheetName + sheetExt)
    })
  }

  function renameSheetName(src, dst) {
    src = path.join(options.thumbnailsDirectory, src)
    dst = path.join(options.thumbnailsDirectory, dst)

    unlinkAndMove(src, dst, removeSpritesDirectory)
  }

  function removeSpritesDirectory(err) {
    if (err) {
      return self.emit('internalError', err)
    }

    fs.remove(spritesDirectory, removeSheetSpec)
  }

  function removeSheetSpec(err) {
    if (err) {
      return self.emit('internalError', err)
    }
    //TODO: ACHTUNG! EVTL WIEDER EINKOMMENTIEREN!
    fs.remove(path.join(options.thumbnailsDirectory, sheetName + '.json'), writeInfo)
  }

  function writeInfo(err) {
    if (err) {
      return self.emit('internalError', err)
    }

    self._writeInfo(paths, jsonMetaDataArray)
  }
}
util.inherits(SpriteSheetWriter, Writer)




module.exports = SpriteSheetWriter
